File: lua/robot-simulator/robot-simulator.lua
CompassPoint = {left = "", right = ""}

function CompassPoint:new (left, right)
    compass = {}
    setmetatable(compass, self)
    self.__index = self
    self.left = left
    self.right = right

    return compass
end

local compass = {
    north = CompassPoint:new("west", "east"),
    south = CompassPoint:new("east", "west"),
    east = CompassPoint:new("north", "south"),
    west = CompassPoint:new("south", "north")
}

Robot = {x = 0, y = 0, heading = ""}

function Robot:new (x, y, heading)
    robot = {}
    setmetatable(robot, self)
    self.__index = self
    self.x = x
    self.y = y
    self.heading = heading

    return robot

end

function Robot:move(commands) 
    commands:gsub(".", function(command)
        if (command == 'A') then
            if (self.heading == "north") then
                self.y = self.y + 1
            elseif (self.heading == "east") then
                self.x = self.x + 1
            elseif (self.heading == "south") then
                self.y = self.y - 1
            else 
                self.x = self.x - 1
            end
        elseif (command == 'R') then
            self.heading = compass[self.heading].right
            -- if (self.heading == "north") then
            --     self.heading = "east"
            -- elseif (self.heading == "east") then
            --     self.heading = "south"
            -- elseif (self.heading == "south") then
            --     self.heading = "west"
            -- else 
            --     self.heading = "north"
            -- end
        elseif (command == 'L') then
            if (self.heading == "north") then
                self.heading = "west"
            elseif (self.heading == "east") then
                self.heading = "north"
            elseif (self.heading == "south") then
                self.heading = "east"
            else 
                self.heading = "south"
            end
        else
            error()
        end

    end)
end

return function(config)
    return Robot:new(config.x, config.y, config.heading)
end

File: solutions/ruby/robot-simulator/1/robot_simulator.rb
class Robot
  def initialize()
    
  end

  @@directions = [:north, :east, :south, :west]

  def orient(direction)

    raise ArgumentError if !@@directions.include?(direction)

    @bearing = direction
  end

  def turn_right()
    bearing_index = @@directions.index(@bearing)

    @bearing = @@directions[(bearing_index + 1) % 4]
  end

  def turn_left()
    bearing_index = @@directions.index(@bearing)

    @bearing = @@directions[(bearing_index - 1) % 4]
  end

  def at(x_coord, y_coord)
    @coordinates = [x_coord, y_coord]
  end

  def advance()
    case @bearing
    when :north
      @coordinates = [@coordinates[0], @coordinates[1] + 1]
    when :south
      @coordinates = [@coordinates[0], @coordinates[1] - 1]
    when :east
      @coordinates = [@coordinates[0] + 1, @coordinates[1]]
    when :west
      @coordinates = [@coordinates[0] - 1, @coordinates[1]]
    end
  end

  attr_reader :bearing, :coordinates

end

class Simulator
  def instructions(command_list)
    commands = []
    command_list.each_char do |command|
      case command
      when 'L'
        commands << :turn_left
      when 'R'
        commands << :turn_right
      when 'A'
        commands << :advance
      end
    end

    commands
  end

  def place(robot, x:, y:, direction:)
    robot.at(x, y)
    robot.orient(direction)
  end

  def evaluate(robot, command_list)
    commands = instructions(command_list)

    commands.each do |command|
      case command
      when :turn_left
        robot.turn_left()
      when :turn_right
        robot.turn_right
      when :advance
        robot.advance()
      end
    end
  end
end